checks.lua

```diff
@@ -10,6 +10,12 @@ local function is_tarantool()
     return _G['_TARANTOOL'] ~= nil
 end
 
+-- Tarantool has its own implementation of startswith().
+-- Function added for compatibility with LuaJIT.
+local function startswith(str, start)
+    return string.sub(str, 1, string.len(start)) == start
+end
```

Comment:
Do you compare performance of our and current functions.

if we need it to check only the first symbol, let's check only it and won't introduce general purpose solution.

Also we could use something like:

```lua
local startswith = string.startswith
if startswith == nil then
local function startswith(str, start)
    return string.sub(str, 1, string.len(start)) == start
end
end
```

Reply:
```lua
tarantool> clock = require('clock')
tarantool> function startswith_ligurio(str, start) return string.sub(str, 1, string.len(start)) == start end
tarantool> do local str = 'foo' local prefix = 'f' local res = {} for i = 1, 10 do table.insert(res, clock.bench(function() for i = 1, 100000000 do str:startswith(prefix) end end)) end return res end
---
- - - 0.39534442
  - - 0.341430259
  - - 0.354928357
  - - 0.341347727
  - - 0.339854973
  - - 0.337526164
  - - 0.336848225
  - - 0.339291624
  - - 0.349172207
  - - 0.34052744
...

tarantool> do local str = 'foo' local prefix = 'f' local res = {} for i = 1, 10 do table.insert(res, clock.bench(function() for i = 1, 100000000 do string.startswith(str, prefix) end end)) end return res end
---
- - - 0.626183345
  - - 0.655380296
  - - 0.33718409
  - - 0.337449185
  - - 0.339688672
  - - 0.341013056
  - - 0.337725581
  - - 0.339039723
  - - 0.33721731
  - - 0.33780096
...

tarantool> do local str = 'foo' local prefix = 'f' local res = {} for i = 1, 10 do table.insert(res, clock.bench(function() for i = 1, 100000000 do startswith_ligurio(str, prefix) end end)) end return res end
---
- - - 0.168886258
  - - 0.169330499
  - - 0.113384887
  - - 0.112382563
  - - 0.112820696
  - - 0.112129318
  - - 0.112136349
  - - 0.111923377
  - - 0.112099052
  - - 0.111906139
...
```

Build options:

```sh
cmake . -DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_BACKTRACE=ON -DENABLE_DIST=ON -DENABLE_FEEDBACK_DAEMON=OFF -DENABLE_BUNDLED_LIBCURL=OFF -DENABLE_FEEDBACK_DAEMON=OFF && make -j
```

Versions:

```sh
$ gcc --version | head -n 1
gcc (Gentoo 11.2.0 p1) 11.2.0
$ git describe 
2.10.0-beta1-331-ga281c7423
```

Reply:
Thanks! It is a valid concern. I have updated the patch and replaced the pure Lua implementation with FFI-based one (the same as used in Tarantool).

Reply:
Pure Lua implementation is faster according to my results, but FFI one was chosen. Is it intentional (and I overlooked something) or by mistake?

Don't know, whether my tests properly reflect a real workload for checks.

Don't know, whether it is really matters. However there are some [observations](https://github.com/tarantool/crud/issues/234) that checks may be visible in a profile in a workload similar to user's one.

Reply:
> Pure Lua implementation is faster according to my results, but FFI one was chosen. Is it intentional (and I overlooked something) or by mistake?

On my machine, the difference is not significant:

```
tarantool> clock = require('clock')                                                                                                     
---                                                                                                                                     
...                                                                                                                                     
                                                                                                                                        
tarantool> function startswith_ligurio(str, start) return string.sub(str, 1, string.len(start)) == start end                            
---                                                                                                                                     
...                                                                                                                                     
                                                                                                                                        
tarantool> do local str = 'foo' local prefix = 'f' local res = {} for i = 1, 10 do table.insert(res, clock.bench(function() for i = 1, 1
00000000 do str:startswith(prefix) end end)) end return res end                                                                         
---                                                                                                                                     
- - - 0.207066875                                                                                                                       
  - - 0.192676831                                                                                                                       
  - - 0.228799382                                                                                                                       
  - - 0.235379259                                                                                                                       
  - - 0.223173765                                                                                                                       
  - - 0.223473087                                                                                                                       
  - - 0.22345024                                                                                                                        
  - - 0.221354018                                                                                                                       
  - - 0.229369831
  - - 0.222026308
...              
tarantool> do local str = 'foo' local prefix = 'f' local res = {} for i = 1, 10 do table.insert(res, clock.bench(function() for i = 1, 1
00000000 do string.startswith(str, prefix) end end)) end return res end
---              
- - - 0.457527155
  - - 0.409576473
  - - 0.202639765
  - - 0.197922334
  - - 0.205304731
  - - 0.214641025
  - - 0.205748042
  - - 0.214150276
  - - 0.21845064
  - - 0.214988928
...
tarantool> do local str = 'foo' local prefix = 'f' local res = {} for i = 1, 10 do table.insert(res, clock.bench(function() for i = 1, 100000000 do startswith_ligurio(str, prefix) end end)) end return res end
---
- - - 0.208166781
  - - 0.195146944
  - - 0.127091714
  - - 0.123113793
  - - 0.121354865
  - - 0.13736582
  - - 0.114352337
  - - 0.121989789
  - - 0.127240625
  - - 0.116291485
...

tarantool> 
```

Tarantool:
```
Tarantool 2.8.3-0-g01023dbc2
Target: Linux-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr -DENABLE_BACKTRACE=ON
Compiler: /usr/bin/cc /usr/lib/ccache/g++
C_FLAGS:-g -O2 -fdebug-prefix-map=/build/tarantool-2.8.3.0=. -specs=/usr/share/dpkg/no-pie-compile.specs -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type
CXX_FLAGS:-g -O2 -fdebug-prefix-map=/build/tarantool-2.8.3.0=. -specs=/usr/share/dpkg/no-pie-compile.specs -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type
```

Reply:
Patch impact with checking different data types (test `perftest.lua`):

![Screenshot from 2022-01-11 15-00-01](https://user-images.githubusercontent.com/1151557/148939104-b4a638f2-5969-4805-8ea3-d034ad0d4876.png)

Results became worse than before patch. I don't know to explain it. It was the same implementation of `startswith()` used in Tarantool.

Data: [Untitled spreadsheet.xlsx](https://github.com/tarantool/checks/files/7846275/Untitled.spreadsheet.xlsx)

Version:
```
Tarantool 2.8.3-0-g01023dbc2
Target: Linux-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr -DENABLE_BACKTRACE=ON
Compiler: /usr/bin/cc /usr/lib/ccache/g++
```

Reply:
Wow, didn't know that there is a performance test here.

While it is on your hands, can you, please, collect `string.sub` results and add relative difference (percents)?

Reply:
Try to use just
```
diff --git a/checks.lua b/checks.lua
index 5cd79d2..e736de4 100644
--- a/checks.lua
+++ b/checks.lua
@@ -1,15 +1,3 @@
-local ffi = require('ffi')
-
-ffi.cdef[[
-    int memcmp(const char *mem1, const char *mem2, size_t num);
-]]
-
-local err_string_arg = "bad argument #%d to '%s' (%s expected, got %s)"
-
-local c_char_ptr     = ffi.typeof('const char *')
-
-local memcmp  = ffi.C.memcmp
-
 local _qualifiers_cache = {
     -- ['?type1|type2'] = {
     --     [1] = 'type1',
@@ -22,57 +10,10 @@ local function is_tarantool()
     return _G['_TARANTOOL'] ~= nil
 end
 
---- Check that string (or substring) starts with given string
--- Optionally restricting the matching with the given offsets
--- @function startswith
--- @string    inp     original string
--- @string    head    the substring to check against
--- @int[opt]  _start  start index of matching boundary
--- @int[opt]  _end    end index of matching boundary
--- @returns           boolean
-local function startswith(inp, head, _start, _end)
-    if type(inp) ~= 'string' then
-        error(err_string_arg:format(1, 'string.startswith', 'string',
-                                    type(inp)), 2)
-    end
-    if type(head) ~= 'string' then
-        error(err_string_arg:format(2, 'string.startswith', 'string',
-                                    type(head)), 2)
-    end
-    if _start ~= nil and type(_start) ~= 'number' then
-        error(err_string_arg:format(3, 'string.startswith', 'integer',
-                                    type(_start)), 2)
-    end
-    if _end ~= nil and type(_end) ~= 'number' then
-        error(err_string_arg:format(4, 'string.startswith', 'integer',
-                                    type(_end)), 2)
-    end
-    -- prepare input arguments (move negative values [offset from the end] to
-    -- positive ones and/or assign default values)
-    local head_len, inp_len = #head, #inp
-    if _start == nil then
-        _start = 1
-    elseif _start < 0 then
-        _start = inp_len + _start + 1
-        if _start < 0 then _start = 0 end
-    end
-    if _end == nil or _end > inp_len then
-        _end = inp_len
-    elseif _end < 0 then
-        _end = inp_len + _end + 1
-        if _end < 0 then _end = 0 end
-    end
-    -- check for degenerate case (interval lesser than input)
-    if head_len == 0 then
-        return true
-    elseif _end - _start + 1 < head_len or _start > _end then
-        return false
-    end
-    _start = _start - 1
-    return memcmp(c_char_ptr(inp) + _start, c_char_ptr(head), head_len) == 0
+local function is_optional(s)
+    return s:byte(1) == 63
 end
 
-
 local function check_string_type(value, expected_type)
     -- 1. Check any value.
     if expected_type == '?' then
@@ -85,7 +26,7 @@ local function check_string_type(value, expected_type)
         qualifier = { optional = false }
 
         for typ in expected_type:gmatch('[^|]+') do
-            if startswith(typ, '?') then
+            if is_optional(typ) then
                 qualifier.optional = true
                 typ = typ:sub(2)
             end

```